File: app/src/main/java/com/mandarin/bcu/androidutil/unit/asynchs/Adder.java
package com.mandarin.bcu.androidutil.unit.asynchs;

import android.app.Activity;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.SystemClock;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.mandarin.bcu.R;
import com.mandarin.bcu.UnitInfo;
import com.mandarin.bcu.androidutil.FilterEntity;
import com.mandarin.bcu.androidutil.StaticStore;
import com.mandarin.bcu.androidutil.unit.Definer;
import com.mandarin.bcu.androidutil.unit.adapters.UnitListAdapter;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Objects;

import common.system.MultiLangCont;
import common.util.pack.Pack;
import common.util.unit.Form;

public class Adder extends AsyncTask<Void, Integer, Void> {
    private final WeakReference<Activity> weakReference;
    private ArrayList<Integer> numbers = new ArrayList<>();

    public Adder(Activity context) {
        this.weakReference = new WeakReference<>(context);
    }

    @Override
    protected void onPreExecute() {
        Activity activity = weakReference.get();

        if(activity == null) return;

        ListView list = activity.findViewById(R.id.unitinflist);
        FloatingActionButton search = activity.findViewById(R.id.animsch);
        TextInputEditText schname = activity.findViewById(R.id.animschname);
        TextInputLayout layout = activity.findViewById(R.id.animschnamel);

        list.setVisibility(View.GONE);
        search.hide();
        schname.setVisibility(View.GONE);
        layout.setVisibility(View.GONE);
    }

    @Override
    protected Void doInBackground(Void... voids) {
        Activity activity = weakReference.get();

        if(activity == null) return null;

        new Definer().define(activity);

        publishProgress(0);

        if (StaticStore.names == null) {
            StaticStore.names = new String[StaticStore.unitnumber];

            for (int i = 0; i < StaticStore.names.length; i++) {
                StaticStore.names[i] = withID(i, MultiLangCont.FNAME.getCont(Pack.def.us.ulist.get(i).forms[0]));
            }
        }

        publishProgress(2);

        return null;
    }

    @Override
    protected void onProgressUpdate(Integer... values) {
        Activity activity = weakReference.get();

        if(activity == null) return;

        TextView ulistst = activity.findViewById(R.id.unitinfst);

        switch(values[0]) {
            case 0:
                ulistst.setText(R.string.unit_list_unitname);
                break;
            case 1:
                ulistst.setText(R.string.unit_list_unitic);
                break;
            case 2:
                ListView list = activity.findViewById(R.id.unitinflist);
                FilterEntity filterEntity;
                TextInputEditText schname = activity.findViewById(R.id.animschname);

                if(Objects.requireNonNull(schname.getText()).toString().isEmpty())
                    filterEntity = new FilterEntity(StaticStore.unitnumber);
                else
                    filterEntity = new FilterEntity(StaticStore.unitnumber,schname.getText().toString());

                numbers = filterEntity.setFilter();
                ArrayList<String> names = new ArrayList<>();

                for (int i : numbers) {
                    names.add(StaticStore.names[i]);
                }

                UnitListAdapter adap = new UnitListAdapter(activity, names.toArray(new String[0]), numbers);
                list.setAdapter(adap);
                list.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                    @Override
                    public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                        Toast.makeText(activity, showName(numbers.get(position)), Toast.LENGTH_SHORT).show();
                        list.setClickable(false);

                        return true;
                    }
                });
                list.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        if (SystemClock.elapsedRealtime() - StaticStore.unitinflistClick < StaticStore.INTERVAL)
                            return;

                        StaticStore.unitinflistClick = SystemClock.elapsedRealtime();

                        Intent result = new Intent(activity, UnitInfo.class);
                        result.putExtra("ID", numbers.get(position));
                        activity.startActivity(result);
                    }
                });

                schname.addTextChangedListener(new TextWatcher() {
                    @Override
                    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                    }

                    @Override
                    public void onTextChanged(CharSequence s, int start, int before, int count) {

                    }

                    @Override
                    public void afterTextChanged(Editable s) {
                        FilterEntity filterEntity = new FilterEntity(StaticStore.unitnumber,s.toString());
                        numbers = filterEntity.setFilter();

                        ArrayList<String> names = new ArrayList<>();

                        for (int i : numbers) {
                            names.add(StaticStore.names[i]);
                        }

                        UnitListAdapter adap = new UnitListAdapter(activity, names.toArray(new String[0]), numbers);
                        list.setAdapter(adap);
                    }
                });

                break;
        }
    }

    @Override
    protected void onPostExecute(Void result) {
        Activity activity = weakReference.get();

        super.onPostExecute(result);

        if(activity == null) return;

        ListView list = activity.findViewById(R.id.unitinflist);
        ProgressBar prog = activity.findViewById(R.id.unitinfprog);
        TextView ulistst = activity.findViewById(R.id.unitinfst);
        FloatingActionButton search = activity.findViewById(R.id.animsch);
        TextInputEditText schname = activity.findViewById(R.id.animschname);
        TextInputLayout layout = activity.findViewById(R.id.animschnamel);
        list.setVisibility(View.VISIBLE);
        prog.setVisibility(View.GONE);
        ulistst.setVisibility(View.GONE);
        search.show();
        schname.setVisibility(View.VISIBLE);
        layout.setVisibility(View.VISIBLE);
    }

    private String number(int num) {
        if (0 <= num && num < 10) {
            return "00" + num;
        } else if (10 <= num && num <= 99) {
            return "0" + num;
        } else {
            return String.valueOf(num);
        }
    }

    private String showName(int location) {
        ArrayList<String> names = new ArrayList<>();

        for(Form f : StaticStore.units.get(location).forms) {
            String name = MultiLangCont.FNAME.getCont(f);
            if(name == null)
                name = "";

            names.add(name);
        }

        StringBuilder result = new StringBuilder(withID(location, names.get(0)));

        for(int i = 1; i < names.size();i++) {
            result.append(" - ").append(names.get(i));
        }

        return result.toString();
    }

    private String withID(int id, String name) {
        String result;
        String names = name;

        if(names == null)
            names = "";

        if(names.equals("")) {
            result = number(id);
        } else {
            result = number(id)+" - "+names;
        }

        return result;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/unit/asynchs/Adder.kt
package com.mandarin.bcu.androidutil.unit.asynchs

import android.app.Activity
import android.os.AsyncTask
import android.os.Parcelable
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.tabs.TabLayout
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import com.mandarin.bcu.androidutil.adapters.MeasureViewPager
import com.mandarin.bcu.androidutil.Definer
import com.mandarin.bcu.androidutil.unit.adapters.UnitListPager
import common.pack.Identifier
import common.pack.PackData
import common.pack.UserProfile
import java.lang.ref.WeakReference

class Adder(context: Activity, private val fm : FragmentManager?) : AsyncTask<Void?, String?, Void?>() {
    private val weakReference: WeakReference<Activity> = WeakReference(context)

    private val image = "3"
    private val castle = "4"
    private val bg = "5"
    private val packext = "6"

    override fun onPreExecute() {
        val activity = weakReference.get() ?: return
        val tab = activity.findViewById<TabLayout>(R.id.unittab)
        val pager = activity.findViewById<MeasureViewPager>(R.id.unitpager)
        val search: FloatingActionButton = activity.findViewById(R.id.animsch)
        val schname: TextInputEditText = activity.findViewById(R.id.animschname)
        val layout: TextInputLayout = activity.findViewById(R.id.animschnamel)
        tab.visibility = View.GONE
        pager.visibility = View.GONE
        search.hide()
        schname.visibility = View.GONE
        layout.visibility = View.GONE
    }

    override fun doInBackground(vararg voids: Void?): Void? {
        val activity = weakReference.get() ?: return null

        Definer.define(activity)
        
        publishProgress("1")

        StaticStore.filterEntityList = BooleanArray(UserProfile.getAllPacks().size)

        publishProgress("2")
        return null
    }

    override fun onProgressUpdate(vararg values: String?) {
        val activity = weakReference.get() ?: return
        val ulistst = activity.findViewById<TextView>(R.id.unitinfst)
        when (values[0]) {
            "1" -> ulistst.setText(R.string.main_pack)
            "2" -> {
                val schname: TextInputEditText = activity.findViewById(R.id.animschname)
                val tab = activity.findViewById<TabLayout>(R.id.unittab)
                val pager = activity.findViewById<MeasureViewPager>(R.id.unitpager)

                if(StaticStore.entityname != "") {
                    schname.setText(StaticStore.entityname)
                }

                schname.addTextChangedListener(object : TextWatcher {
                    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                    override fun afterTextChanged(s: Editable) {
                        StaticStore.entityname = s.toString()

                        for(i in StaticStore.filterEntityList.indices) {
                            StaticStore.filterEntityList[i] = true
                        }
                    }
                })

                fm ?: return

                pager.removeAllViewsInLayout()
                pager.adapter = UnitListTab(fm)
                pager.offscreenPageLimit = getExistingUnit()
                pager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tab))

                tab.setupWithViewPager(pager)
            }

            image -> {
                val name = activity.getString(R.string.main_pack_img)+ (values[1] ?: "")

                ulistst.text = name
            }

            bg -> {
                val name = activity.getString(R.string.main_pack_bg) + (values[1] ?: "")

                ulistst.text = name
            }

            castle -> {
                val name = activity.getString(R.string.main_pack_castle) + (values[1] ?: "")

                ulistst.text = name
            }

            packext -> {
                val name = activity.getString(R.string.main_pack_ext)+ (values[1] ?: "")

                ulistst.text = name
            }
        }
    }

    override fun onPostExecute(result: Void?) {
        val activity = weakReference.get()
        super.onPostExecute(result)
        if (activity == null) return
        val tab = activity.findViewById<TabLayout>(R.id.unittab)
        val pager = activity.findViewById<MeasureViewPager>(R.id.unitpager)
        val prog = activity.findViewById<ProgressBar>(R.id.unitinfprog)
        val search: FloatingActionButton = activity.findViewById(R.id.animsch)
        val schname: TextInputEditText = activity.findViewById(R.id.animschname)
        val layout: TextInputLayout = activity.findViewById(R.id.animschnamel)
        val loadt = activity.findViewById<TextView>(R.id.unitinfst)
        loadt.visibility = View.GONE
        if(UserProfile.getAllPacks().size != 1) {
            tab.visibility = View.VISIBLE
        }
        pager.visibility = View.VISIBLE
        prog.visibility = View.GONE
        search.show()
        schname.visibility = View.VISIBLE
        layout.visibility = View.VISIBLE
    }

    private fun getExistingUnit() : Int {
        var i = 0

        for(p in UserProfile.getAllPacks()) {
            if(p.units.list.isNotEmpty())
                i++
        }

        return i
    }

    inner class UnitListTab internal constructor(fm: FragmentManager) : FragmentStatePagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
        private val keys: ArrayList<String>

        init {
            val lit = fm.fragments
            val trans = fm.beginTransaction()

            for(f in lit) {
                trans.remove(f)
            }

            trans.commitAllowingStateLoss()

            keys = getExistingPack()
        }

        override fun getItem(position: Int): Fragment {
            return UnitListPager.newInstance(keys[position], position)
        }

        override fun getCount(): Int {
            return keys.size
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return if(position == 0) {
                "Default"
            } else {
                val pack = PackData.getPack(keys[position])

                if(pack == null) {
                    keys[position]
                }

                val name = when (pack) {
                    is PackData.DefPack -> {
                        weakReference.get()?.getString(R.string.pack_default) ?: ""
                    }
                    is PackData.UserPack -> {
                        StaticStore.getPackName(pack.desc.id)
                    }
                    else -> {
                        ""
                    }
                }

                if(name.isEmpty()) {
                    keys[position]
                } else {
                    name
                }
            }
        }

        override fun saveState(): Parcelable? {
            return null
        }

        private fun getExistingPack() : ArrayList<String> {
            val packs = UserProfile.getAllPacks()

            val res = ArrayList<String>()

            for(p in packs) {
                if(p is PackData.DefPack) {
                    res.add(Identifier.DEF)
                } else if(p is PackData.UserPack) {
                    res.add(p.desc.id)
                }
            }

            return res
        }
    }
}